File: go/core/conformance.go
package core

// Projects pipeline output into the plain Value wire artifacts the conformance
// goldens pin — port of conformance.ts / conformance.rs. This file holds the
// var-doc projection; registry/plan/trace projections are added in later stages.

func obj(pairs ...[2]any) Value {
	m := make(map[string]Value, len(pairs))
	for _, p := range pairs {
		m[p[0].(string)] = p[1].(Value)
	}
	return Value{Kind: KindMap, Map: m}
}

func kv(k string, v Value) [2]any { return [2]any{k, v} }

func vint(n int) Value { return IntValue(int64(n)) }

// ToVarDocArtifact projects a parsed VarDoc to the var-doc wire artifact.
func ToVarDocArtifact(doc VarDoc) Value {
	examples := make([]Value, len(doc.Examples))
	for i := range doc.Examples {
		examples[i] = exampleValue(doc.Examples[i])
	}
	orphans := make([]Value, len(doc.OrphanAttachments))
	for i, o := range doc.OrphanAttachments {
		orphans[i] = tableOrFenceValue(o)
	}
	return obj(
		kv("path", StrValue(doc.Path)),
		kv("examples", ListOf(examples)),
		kv("orphanAttachments", ListOf(orphans)),
	)
}

func spanValue(s Span) Value {
	return obj(
		kv("startOffset", vint(s.StartOffset)),
		kv("endOffset", vint(s.EndOffset)),
		kv("startLine", vint(s.StartLine)),
		kv("startCol", vint(s.StartCol)),
		kv("endLine", vint(s.EndLine)),
		kv("endCol", vint(s.EndCol)),
	)
}

func segmentOffsetValue(o SegmentOffset) Value {
	return obj(
		kv("textOffset", vint(o.TextOffset)),
		kv("sourceOffset", vint(o.SourceOffset)),
	)
}

func segmentMapValue(m []SegmentOffset) Value {
	out := make([]Value, len(m))
	for i, o := range m {
		out[i] = segmentOffsetValue(o)
	}
	return ListOf(out)
}

func rowValue(r Row) Value {
	cells := make([]Value, len(r.Cells))
	for i, c := range r.Cells {
		cells[i] = StrValue(c)
	}
	spans := make([]Value, len(r.CellSpans))
	for i, s := range r.CellSpans {
		spans[i] = spanValue(s)
	}
	return obj(
		kv("cells", ListOf(cells)),
		kv("cellSpans", ListOf(spans)),
		kv("span", spanValue(r.Span)),
	)
}

func tableValue(t Table) Value {
	rows := make([]Value, len(t.Rows))
	for i := range t.Rows {
		rows[i] = rowValue(t.Rows[i])
	}
	return obj(
		kv("kind", StrValue("table")),
		kv("span", spanValue(t.Span)),
		kv("header", rowValue(t.Header)),
		kv("rows", ListOf(rows)),
	)
}

func fenceValue(f Fence) Value {
	return obj(
		kv("kind", StrValue("fence")),
		kv("span", spanValue(f.Span)),
		kv("info", StrValue(f.Info)),
		kv("body", StrValue(f.Body)),
		kv("bodySpan", spanValue(f.BodySpan)),
	)
}

func headingValue(h Heading) Value {
	return obj(
		kv("kind", StrValue("heading")),
		kv("level", vint(h.Level)),
		kv("text", StrValue(h.Text)),
		kv("span", spanValue(h.Span)),
	)
}

func paragraphValue(p Paragraph) Value {
	return obj(
		kv("kind", StrValue("paragraph")),
		kv("text", StrValue(p.Text)),
		kv("span", spanValue(p.Span)),
		kv("segmentMap", segmentMapValue(p.SegmentMap)),
	)
}

func listItemValue(l ListItem) Value {
	return obj(
		kv("kind", StrValue("list_item")),
		kv("text", StrValue(l.Text)),
		kv("span", spanValue(l.Span)),
		kv("segmentMap", segmentMapValue(l.SegmentMap)),
		kv("ordered", BoolValue(l.Ordered)),
		kv("markerSpan", spanValue(l.MarkerSpan)),
	)
}

func blockquoteValue(b Blockquote) Value {
	return obj(
		kv("kind", StrValue("blockquote")),
		kv("text", StrValue(b.Text)),
		kv("span", spanValue(b.Span)),
		kv("segmentMap", segmentMapValue(b.SegmentMap)),
	)
}

func thematicBreakValue(t ThematicBreak) Value {
	return obj(
		kv("kind", StrValue("thematic_break")),
		kv("span", spanValue(t.Span)),
	)
}

func blockValue(b Block) Value {
	switch v := b.(type) {
	case Heading:
		return headingValue(v)
	case Paragraph:
		return paragraphValue(v)
	case ListItem:
		return listItemValue(v)
	case Blockquote:
		return blockquoteValue(v)
	case Table:
		return tableValue(v)
	case Fence:
		return fenceValue(v)
	case ThematicBreak:
		return thematicBreakValue(v)
	}
	panic("unknown block type")
}

func tableOrFenceValue(b Block) Value {
	switch v := b.(type) {
	case Table:
		return tableValue(v)
	case Fence:
		return fenceValue(v)
	}
	panic("orphan attachment must be table or fence")
}

// BundleArtifacts are all four projected wire artifacts for one bundle.
type BundleArtifacts struct {
	VarDoc   Value
	Registry Value
	Plan     Value
	Trace    Value
}

// ToFailureArtifact projects a caught step failure to the FailureArtifact wire
// shape. A nil failure falls through to "thrown".
func ToFailureArtifact(failure *StepFailure, matchSpan Span) Value {
	line := matchSpan.StartLine
	anchorSpan := matchSpan
	if failure != nil {
		anchorSpan = anchor(failure.Error, matchSpan)
	}
	anchorVal := spanValue(anchorSpan)

	if failure == nil {
		return kindLineAnchor("thrown", line, anchorVal)
	}
	switch failure.Error.Kind {
	case SECellMismatch:
		var failing []Value
		for _, c := range failure.Error.Cells {
			if !c.Ok {
				failing = append(failing, failureCell(c))
			}
		}
		return obj(
			kv("kind", StrValue("cell-mismatch")),
			kv("line", vint(line)),
			kv("anchor", anchorVal),
			kv("cells", ListOf(failing)),
		)
	case SEDocStringMismatch:
		d := obj(
			kv("expected", StrValue(failure.Error.DocDiff.Expected)),
			kv("actual", StrValue(failure.Error.DocDiff.Actual)),
			kv("span", spanValue(failure.Error.DocDiff.Span)),
		)
		return obj(
			kv("kind", StrValue("doc-string-mismatch")),
			kv("line", vint(line)),
			kv("anchor", anchorVal),
			kv("diff", d),
		)
	case SEReturnShape:
		return kindLineAnchor("return-shape", line, anchorVal)
	case SEUnexpectedPass:
		return kindLineAnchor("unexpected-pass", line, anchorVal)
	default:
		return kindLineAnchor("thrown", line, anchorVal)
	}
}

func failureCell(c CellDiff) Value {
	return obj(
		kv("column", StrValue(c.Column)),
		kv("expected", StrValue(c.Expected)),
		kv("actual", StrValue(c.Actual)),
		kv("span", spanValue(c.Span)),
	)
}

func kindLineAnchor(kind string, line int, anchorVal Value) Value {
	return obj(
		kv("kind", StrValue(kind)),
		kv("line", vint(line)),
		kv("anchor", anchorVal),
	)
}

// fileStem recovers the cross-language-shared step-file stem (strip the last
// extension), e.g. numerals.steps.go → numerals.steps.
func fileStem(path string) string {
	base := path
	for i := len(path) - 1; i >= 0; i-- {
		if path[i] == '/' || path[i] == '\\' {
			base = path[i+1:]
			break
		}
	}
	dot := -1
	for i := len(base) - 1; i >= 0; i-- {
		if base[i] == '.' {
			dot = i
			break
		}
	}
	if dot > 0 {
		return base[:dot]
	}
	return base
}

// RunConformance runs one bundle end-to-end: plan, execute (recording
// observations), and project all four wire artifacts. Port of runConformance.
func RunConformance(doc VarDoc, registry Registry, stateFactory func() Value) BundleArtifacts {
	execution := Plan(doc, registry)

	observed := map[int][]StepObservation{}
	ports := ExecutePorts{
		Reporter:      func(Diagnostic) {},
		CreateContext: func(string) Value { return stateFactory() },
		Observer: func(o StepObservation) {
			observed[o.ExampleIndex] = append(observed[o.ExampleIndex], o)
		},
	}

	queue := CollectExamples(execution, ports)
	traceExamples := make([]Value, 0, len(queue))
	for k, queued := range queue {
		outcome := "pass"
		if queued.Run() != nil {
			outcome = "fail"
		}

		planned := execution.Examples[k]
		obs := observed[k]

		steps := make([]Value, 0, len(planned.Steps))
		for i := range planned.Steps {
			step := planned.Steps[i]
			ordinal := i + 1
			// Prefer the first "fail" observation for this ordinal; else the last.
			var chosen *StepObservation
			for j := range obs {
				if obs[j].Ordinal != ordinal {
					continue
				}
				chosen = &obs[j]
				if obs[j].Outcome == OutcomeFail {
					break
				}
			}
			stepOutcome := "skipped"
			if chosen != nil {
				stepOutcome = chosen.Outcome.String()
			}

			contextKey := obj(
				kv("exampleName", StrValue(queued.Name)),
				kv("stepFile", StrValue(fileStem(step.StepDef.ExpressionSourceFile))),
			)
			m := map[string]Value{
				"exampleName":       StrValue(queued.Name),
				"ordinal":           vint(ordinal),
				"stepText":          StrValue(step.Text),
				"matchedExpression": StrValue(step.StepDef.Expression),
				"contextKey":        contextKey,
				"outcome":           StrValue(stepOutcome),
			}
			if stepOutcome == "fail" {
				var failure *StepFailure
				if chosen != nil {
					failure = chosen.Error
				}
				m["failure"] = ToFailureArtifact(failure, step.MatchSpan)
			}
			steps = append(steps, Value{Kind: KindMap, Map: m})
		}

		traceExamples = append(traceExamples, obj(
			kv("name", StrValue(queued.Name)),
			kv("outcome", StrValue(outcome)),
			kv("steps", ListOf(steps)),
		))
	}

	trace := obj(kv("examples", ListOf(traceExamples)))

	return BundleArtifacts{
		VarDoc:   ToVarDocArtifact(doc),
		Registry: ToRegistryArtifact(registry),
		Plan:     ToPlanArtifact(execution),
		Trace:    trace,
	}
}

// ToRegistryArtifact projects a Registry to the registry wire artifact.
func ToRegistryArtifact(registry Registry) Value {
	steps := make([]Value, len(registry.Steps))
	for i, s := range registry.Steps {
		names := ParameterTypeNames(s.Expression)
		nameVals := make([]Value, len(names))
		for j, n := range names {
			nameVals[j] = StrValue(n)
		}
		steps[i] = obj(
			kv("expression", StrValue(s.Expression)),
			kv("parameterTypeNames", ListOf(nameVals)),
		)
	}
	pts := make([]Value, len(registry.CustomParameterTypes))
	for i, p := range registry.CustomParameterTypes {
		pts[i] = obj(
			kv("name", StrValue(p.Name)),
			kv("regexp", StrValue(p.Regexp)),
		)
	}
	return obj(
		kv("steps", ListOf(steps)),
		kv("parameterTypes", ListOf(pts)),
	)
}

// ToPlanArtifact projects an ExecutionPlan to the plan wire artifact.
func ToPlanArtifact(plan ExecutionPlan) Value {
	source := plan.VarDoc.Source
	examples := make([]Value, len(plan.Examples))
	for i := range plan.Examples {
		examples[i] = plannedExampleValue(source, plan.Examples[i])
	}
	diags := make([]Value, len(plan.Diagnostics))
	for i, d := range plan.Diagnostics {
		diags[i] = diagnosticValue(d)
	}
	return obj(
		kv("examples", ListOf(examples)),
		kv("diagnostics", ListOf(diags)),
	)
}

func plannedExampleValue(source string, ex PlannedExample) Value {
	scope := make([]Value, len(ex.ScopeStack))
	for i, s := range ex.ScopeStack {
		scope[i] = StrValue(s)
	}
	outcome := "pass"
	if ex.ExpectedOutcome != nil {
		outcome = *ex.ExpectedOutcome
	}
	m := map[string]Value{
		"name":            StrValue(ex.Name),
		"scopeStack":      ListOf(scope),
		"span":            spanValue(ex.Span),
		"expectedOutcome": StrValue(outcome),
	}
	if ex.ExpectedErrorMessage != nil {
		m["expectedErrorMessage"] = StrValue(*ex.ExpectedErrorMessage)
	}
	steps := make([]Value, len(ex.Steps))
	for i := range ex.Steps {
		steps[i] = plannedStepValue(source, ex.Steps[i])
	}
	m["steps"] = ListOf(steps)
	return Value{Kind: KindMap, Map: m}
}

func plannedStepValue(source string, step PlannedStep) Value {
	paramNames := ParameterTypeNames(step.StepDef.Expression)
	args := make([]Value, len(step.ParamSpans))
	for i, ps := range step.ParamSpans {
		var paramType Value
		if i < len(paramNames) {
			paramType = StrValue(paramNames[i])
		} else {
			paramType = NullValue
		}
		args[i] = obj(
			kv("value", StrValue(utf16Slice(source, ps.StartOffset, ps.EndOffset))),
			kv("parameterType", paramType),
		)
	}
	spans := make([]Value, len(step.ParamSpans))
	for i, s := range step.ParamSpans {
		spans[i] = spanValue(s)
	}
	m := map[string]Value{
		"text":              StrValue(step.Text),
		"matchSpan":         spanValue(step.MatchSpan),
		"paramSpans":        ListOf(spans),
		"matchedExpression": StrValue(step.StepDef.Expression),
		"args":              ListOf(args),
	}
	if step.DataTable != nil {
		m["dataTable"] = tableValue(*step.DataTable)
	}
	if step.DocString != nil {
		m["docString"] = docStringValue(*step.DocString)
	}
	return Value{Kind: KindMap, Map: m}
}

func docStringValue(f Fence) Value {
	return obj(
		kv("content", StrValue(f.Body)),
		kv("contentType", StrValue(f.Info)),
		kv("span", spanValue(f.BodySpan)),
	)
}

func diagnosticValue(d Diagnostic) Value {
	return obj(
		kv("code", StrValue(diagnosticCodeString(d.Code))),
		kv("severity", StrValue(severityString(d.Severity))),
		kv("span", spanValue(d.Span)),
	)
}

func diagnosticCodeString(code DiagnosticCode) string {
	switch code {
	case CodeAmbiguousMatch:
		return "ambiguous-match"
	case CodeErrorFenceWithoutStep:
		return "error-fence-without-step"
	case CodeDrift:
		return "drift"
	}
	return ""
}

func severityString(s Severity) string {
	switch s {
	case SeverityError:
		return "error"
	case SeverityWarning:
		return "warning"
	case SeverityInfo:
		return "info"
	}
	return ""
}

func exampleValue(e Example) Value {
	scope := make([]Value, len(e.ScopeStack))
	for i, s := range e.ScopeStack {
		scope[i] = StrValue(s)
	}
	body := make([]Value, len(e.Body))
	for i, blk := range e.Body {
		body[i] = blockValue(blk)
	}
	return obj(
		kv("scopeStack", ListOf(scope)),
		kv("span", spanValue(e.Span)),
		kv("body", ListOf(body)),
	)
}

File: rust/core/src/conformance.rs
//! Projects pipeline output into the plain [`Value`] wire artifacts the
//! conformance goldens pin — port of `conformance.ts` / `Conformance.java`.
//! Covers all four projections (var-doc, registry, plan, trace).

use crate::ast::{
    Block, Blockquote, Example, Fence, Heading, ListItem, Paragraph, Row, SegmentOffset, Table,
    TableOrFence, ThematicBreak, VarDoc,
};
use crate::diagnostics::{Diagnostic, DiagnosticCode, Severity};
use crate::error::{StepError, StepFailure};
use crate::execute::{ExecutePorts, StepObservation, StepOutcome, collect_examples};
use crate::offsets::utf16_slice;
use crate::plan::{ExecutionPlan, PlannedExample, PlannedStep, plan};
use crate::registry::Registry;
use crate::span::Span;
use crate::value::Value;
use std::cell::RefCell;
use std::collections::{BTreeMap, HashMap};
use std::rc::Rc;

pub use crate::expression::parameter_type_names;

/// All four projected wire artifacts for one bundle.
pub struct BundleArtifacts {
    pub var_doc: Value,
    pub registry: Value,
    pub plan: Value,
    pub trace: Value,
}

fn obj(pairs: Vec<(&str, Value)>) -> Value {
    let mut m = BTreeMap::new();
    for (k, v) in pairs {
        m.insert(k.to_string(), v);
    }
    Value::Map(m)
}

fn vint(n: usize) -> Value {
    Value::Int(n as i64)
}

// -----------------------------------------------------------------------------
// var-doc projection
// -----------------------------------------------------------------------------

/// Projects a parsed [`VarDoc`] to the var-doc wire artifact.
pub fn to_var_doc_artifact(doc: &VarDoc) -> Value {
    obj(vec![
        ("path", Value::from(doc.path.as_str())),
        (
            "examples",
            Value::List(doc.examples.iter().map(example).collect()),
        ),
        (
            "orphanAttachments",
            Value::List(doc.orphan_attachments.iter().map(table_or_fence).collect()),
        ),
    ])
}

fn span(s: Span) -> Value {
    obj(vec![
        ("startOffset", vint(s.start_offset)),
        ("endOffset", vint(s.end_offset)),
        ("startLine", vint(s.start_line)),
        ("startCol", vint(s.start_col)),
        ("endLine", vint(s.end_line)),
        ("endCol", vint(s.end_col)),
    ])
}

fn segment_offset(o: &SegmentOffset) -> Value {
    obj(vec![
        ("textOffset", vint(o.text_offset)),
        ("sourceOffset", vint(o.source_offset)),
    ])
}

fn segment_map(map: &[SegmentOffset]) -> Value {
    Value::List(map.iter().map(segment_offset).collect())
}

fn row(r: &Row) -> Value {
    obj(vec![
        (
            "cells",
            Value::List(r.cells.iter().map(|c| Value::from(c.as_str())).collect()),
        ),
        (
            "cellSpans",
            Value::List(r.cell_spans.iter().map(|s| span(*s)).collect()),
        ),
        ("span", span(r.span)),
    ])
}

fn table(t: &Table) -> Value {
    obj(vec![
        ("kind", Value::from("table")),
        ("span", span(t.span)),
        ("header", row(&t.header)),
        ("rows", Value::List(t.rows.iter().map(row).collect())),
    ])
}

fn fence(f: &Fence) -> Value {
    obj(vec![
        ("kind", Value::from("fence")),
        ("span", span(f.span)),
        ("info", Value::from(f.info.as_str())),
        ("body", Value::from(f.body.as_str())),
        ("bodySpan", span(f.body_span)),
    ])
}

fn heading(h: &Heading) -> Value {
    obj(vec![
        ("kind", Value::from("heading")),
        ("level", vint(h.level)),
        ("text", Value::from(h.text.as_str())),
        ("span", span(h.span)),
    ])
}

fn paragraph(p: &Paragraph) -> Value {
    obj(vec![
        ("kind", Value::from("paragraph")),
        ("text", Value::from(p.text.as_str())),
        ("span", span(p.span)),
        ("segmentMap", segment_map(&p.segment_map)),
    ])
}

fn list_item(l: &ListItem) -> Value {
    obj(vec![
        ("kind", Value::from("list_item")),
        ("text", Value::from(l.text.as_str())),
        ("span", span(l.span)),
        ("segmentMap", segment_map(&l.segment_map)),
        ("ordered", Value::Bool(l.ordered)),
        ("markerSpan", span(l.marker_span)),
    ])
}

fn blockquote(b: &Blockquote) -> Value {
    obj(vec![
        ("kind", Value::from("blockquote")),
        ("text", Value::from(b.text.as_str())),
        ("span", span(b.span)),
        ("segmentMap", segment_map(&b.segment_map)),
    ])
}

fn thematic_break(t: &ThematicBreak) -> Value {
    obj(vec![
        ("kind", Value::from("thematic_break")),
        ("span", span(t.span)),
    ])
}

fn block(b: &Block) -> Value {
    match b {
        Block::Heading(h) => heading(h),
        Block::Paragraph(p) => paragraph(p),
        Block::ListItem(l) => list_item(l),
        Block::Blockquote(b) => blockquote(b),
        Block::Table(t) => table(t),
        Block::Fence(f) => fence(f),
        Block::ThematicBreak(t) => thematic_break(t),
    }
}

fn table_or_fence(tf: &TableOrFence) -> Value {
    match tf {
        TableOrFence::Table(t) => table(t),
        TableOrFence::Fence(f) => fence(f),
    }
}

fn example(e: &Example) -> Value {
    obj(vec![
        (
            "scopeStack",
            Value::List(
                e.scope_stack
                    .iter()
                    .map(|s| Value::from(s.as_str()))
                    .collect(),
            ),
        ),
        ("span", span(e.span)),
        ("body", Value::List(e.body.iter().map(block).collect())),
    ])
}

// -----------------------------------------------------------------------------
// registry projection
// -----------------------------------------------------------------------------

/// Projects a [`Registry`] to the registry wire artifact.
pub fn to_registry_artifact(registry: &Registry) -> Value {
    let steps: Vec<Value> = registry
        .steps
        .iter()
        .map(|s| {
            obj(vec![
                ("expression", Value::from(s.expression.as_str())),
                (
                    "parameterTypeNames",
                    Value::List(
                        parameter_type_names(&s.expression)
                            .into_iter()
                            .map(Value::from)
                            .collect(),
                    ),
                ),
            ])
        })
        .collect();
    let parameter_types: Vec<Value> = registry
        .custom_parameter_types
        .iter()
        .map(|p| {
            obj(vec![
                ("name", Value::from(p.name.as_str())),
                ("regexp", Value::from(p.regexp.as_str())),
            ])
        })
        .collect();
    obj(vec![
        ("steps", Value::List(steps)),
        ("parameterTypes", Value::List(parameter_types)),
    ])
}

// -----------------------------------------------------------------------------
// plan projection
// -----------------------------------------------------------------------------

/// Projects an [`ExecutionPlan`] to the plan wire artifact.
pub fn to_plan_artifact(plan: &ExecutionPlan) -> Value {
    let source = &plan.var_doc.source;
    obj(vec![
        (
            "examples",
            Value::List(
                plan.examples
                    .iter()
                    .map(|ex| planned_example(source, ex))
                    .collect(),
            ),
        ),
        (
            "diagnostics",
            Value::List(plan.diagnostics.iter().map(diagnostic).collect()),
        ),
    ])
}

fn planned_example(source: &str, ex: &PlannedExample) -> Value {
    let mut pairs = vec![
        ("name", Value::from(ex.name.as_str())),
        (
            "scopeStack",
            Value::List(
                ex.scope_stack
                    .iter()
                    .map(|s| Value::from(s.as_str()))
                    .collect(),
            ),
        ),
        ("span", span(ex.span)),
        (
            "expectedOutcome",
            Value::from(ex.expected_outcome.as_deref().unwrap_or("pass")),
        ),
    ];
    if let Some(msg) = &ex.expected_error_message {
        pairs.push(("expectedErrorMessage", Value::from(msg.as_str())));
    }
    pairs.push((
        "steps",
        Value::List(ex.steps.iter().map(|s| planned_step(source, s)).collect()),
    ));
    obj(pairs)
}

fn planned_step(source: &str, step: &PlannedStep) -> Value {
    let param_names = parameter_type_names(&step.step_def.expression);
    let args: Vec<Value> = step
        .param_spans
        .iter()
        .enumerate()
        .map(|(i, ps)| {
            obj(vec![
                (
                    "value",
                    Value::from(utf16_slice(source, ps.start_offset, ps.end_offset)),
                ),
                (
                    "parameterType",
                    param_names
                        .get(i)
                        .map_or(Value::Null, |n| Value::from(n.as_str())),
                ),
            ])
        })
        .collect();

    let mut pairs = vec![
        ("text", Value::from(step.text.as_str())),
        ("matchSpan", span(step.match_span)),
        (
            "paramSpans",
            Value::List(step.param_spans.iter().map(|s| span(*s)).collect()),
        ),
        (
            "matchedExpression",
            Value::from(step.step_def.expression.as_str()),
        ),
        ("args", Value::List(args)),
    ];
    if let Some(t) = &step.data_table {
        pairs.push(("dataTable", table(t)));
    }
    if let Some(f) = &step.doc_string {
        pairs.push(("docString", doc_string(f)));
    }
    obj(pairs)
}

fn doc_string(f: &Fence) -> Value {
    obj(vec![
        ("content", Value::from(f.body.as_str())),
        ("contentType", Value::from(f.info.as_str())),
        ("span", span(f.body_span)),
    ])
}

fn diagnostic(d: &Diagnostic) -> Value {
    obj(vec![
        ("code", Value::from(diagnostic_code(d.code))),
        ("severity", Value::from(severity(d.severity))),
        ("span", span(d.span)),
    ])
}

fn diagnostic_code(code: DiagnosticCode) -> &'static str {
    match code {
        DiagnosticCode::AmbiguousMatch => "ambiguous-match",
        DiagnosticCode::ErrorFenceWithoutStep => "error-fence-without-step",
        DiagnosticCode::Drift => "drift",
    }
}

fn severity(s: Severity) -> &'static str {
    match s {
        Severity::Error => "error",
        Severity::Warning => "warning",
        Severity::Info => "info",
    }
}

// -----------------------------------------------------------------------------
// trace projection
// -----------------------------------------------------------------------------

/// Projects a caught step failure to the `FailureArtifact` wire shape. `None`
/// error falls through to `"thrown"`.
pub fn to_failure_artifact(failure: Option<&StepFailure>, match_span: Span) -> Value {
    let line = match_span.start_line;
    let anchor_span = match failure {
        Some(f) => crate::failure_anchor::anchor(&f.error, match_span),
        None => match_span,
    };
    let anchor = span(anchor_span);

    match failure.map(|f| &f.error) {
        Some(StepError::CellMismatch(cells)) => {
            let failing: Vec<Value> = cells.iter().filter(|c| !c.ok).map(failure_cell).collect();
            obj(vec![
                ("kind", Value::from("cell-mismatch")),
                ("line", vint(line)),
                ("anchor", anchor),
                ("cells", Value::List(failing)),
            ])
        }
        Some(StepError::DocStringMismatch(diff)) => {
            let d = obj(vec![
                ("expected", Value::from(diff.expected.as_str())),
                ("actual", Value::from(diff.actual.as_str())),
                ("span", span(diff.span)),
            ]);
            obj(vec![
                ("kind", Value::from("doc-string-mismatch")),
                ("line", vint(line)),
                ("anchor", anchor),
                ("diff", d),
            ])
        }
        Some(StepError::ReturnShape(_)) => kind_line_anchor("return-shape", line, anchor),
        Some(StepError::UnexpectedPass) => kind_line_anchor("unexpected-pass", line, anchor),
        _ => kind_line_anchor("thrown", line, anchor),
    }
}

fn failure_cell(c: &crate::cell_diff::CellDiff) -> Value {
    obj(vec![
        ("column", Value::from(c.column.as_str())),
        ("expected", Value::from(c.expected.as_str())),
        ("actual", Value::from(c.actual.as_str())),
        ("span", span(c.span)),
    ])
}

fn kind_line_anchor(kind: &str, line: usize, anchor: Value) -> Value {
    obj(vec![
        ("kind", Value::from(kind)),
        ("line", vint(line)),
        ("anchor", anchor),
    ])
}

/// Recovers the cross-language-shared step-file stem (strip the last extension),
/// e.g. `numerals.steps.rs` → `numerals.steps`.
fn file_stem(path: &str) -> String {
    let base = path.rsplit(['/', '\\']).next().unwrap_or(path);
    match base.rfind('.') {
        Some(dot) if dot > 0 => base[..dot].to_string(),
        _ => base.to_string(),
    }
}

/// Runs one bundle end-to-end: plan, execute (recording observations), and
/// project all four wire artifacts. Port of `runConformance`.
pub fn run_conformance(
    doc: &VarDoc,
    registry: &Registry,
    context_factory: &dyn Fn() -> Value,
) -> BundleArtifacts {
    let execution = plan(doc, registry);

    let observed: Rc<RefCell<HashMap<usize, Vec<StepObservation>>>> =
        Rc::new(RefCell::new(HashMap::new()));
    let observed_writer = observed.clone();
    let ports = ExecutePorts {
        reporter: Box::new(|_| {}),
        create_context: Some(Box::new(|_| context_factory())),
        observer: Some(Box::new(move |o: StepObservation| {
            observed_writer
                .borrow_mut()
                .entry(o.example_index)
                .or_default()
                .push(o);
        })),
    };

    let queue = collect_examples(&execution, &ports);
    let mut trace_examples = Vec::with_capacity(queue.len());
    for (k, queued) in queue.iter().enumerate() {
        let outcome = if queued.run().is_err() {
            "fail"
        } else {
            "pass"
        };

        let planned = &execution.examples[k];
        let empty = Vec::new();
        let obs_map = observed.borrow();
        let obs = obs_map.get(&k).unwrap_or(&empty);

        let mut steps = Vec::with_capacity(planned.steps.len());
        for (i, step) in planned.steps.iter().enumerate() {
            let ordinal = i + 1;
            // Prefer the first "fail" observation for this ordinal; else the last.
            let mut chosen: Option<&StepObservation> = None;
            for o in obs {
                if o.ordinal != ordinal {
                    continue;
                }
                chosen = Some(o);
                if o.outcome == StepOutcome::Fail {
                    break;
                }
            }
            let step_outcome = chosen.map_or("skipped", |o| o.outcome.as_str());

            let context_key = obj(vec![
                ("exampleName", Value::from(queued.name.as_str())),
                (
                    "stepFile",
                    Value::from(file_stem(&step.step_def.expression_source_file).as_str()),
                ),
            ]);
            let mut step_pairs = vec![
                ("exampleName", Value::from(queued.name.as_str())),
                ("ordinal", vint(ordinal)),
                ("stepText", Value::from(step.text.as_str())),
                (
                    "matchedExpression",
                    Value::from(step.step_def.expression.as_str()),
                ),
                ("contextKey", context_key),
                ("outcome", Value::from(step_outcome)),
            ];
            if step_outcome == "fail" {
                let failure = chosen.and_then(|o| o.error.as_ref());
                step_pairs.push(("failure", to_failure_artifact(failure, step.match_span)));
            }
            steps.push(obj(step_pairs));
        }

        trace_examples.push(obj(vec![
            ("name", Value::from(queued.name.as_str())),
            ("outcome", Value::from(outcome)),
            ("steps", Value::List(steps)),
        ]));
    }

    let trace = obj(vec![("examples", Value::List(trace_examples))]);

    BundleArtifacts {
        var_doc: to_var_doc_artifact(doc),
        registry: to_registry_artifact(registry),
        plan: to_plan_artifact(&execution),
        trace,
    }
}
